File: agent-packs/eval-pilot/engine/src/evalpilot/metrics.py
"""Numeric metric results with JSONL history and trend/regression compare.

This is the half of the result model that the binary rubric cannot express:
a *number* you want to track across runs (latency, token cost, a judge
score, a coverage percentage, an artifact word count, …). Each recorded
value is appended as one JSON line to a **committed** history file:

    <eval_root>/_metrics/<metric-slug>/history.jsonl

Because the history is plain JSONL committed in the target repo, trends are
git-diffable and portable — no database, no external service.

Recording a metric also **compares it to a baseline** and flags
regressions. The baseline is resolved by strategy:

* ``"last"``        — the most recent prior recorded value (default).
* ``"rolling_mean"``— mean of the last ``window`` prior values.
* ``"best"``        — the best prior value given ``direction``.
* ``"pinned"``      — an explicit ``baseline=`` value you pass in.

``direction`` decides what "worse" means:

* ``"higher_is_better"`` — a *drop* beyond tolerance is a regression.
* ``"lower_is_better"``  — a *rise* beyond tolerance is a regression.
* ``"neutral"``          — never flagged; recorded for information only.

Tolerance is ``tolerance`` (absolute) and/or ``tolerance_pct`` (fraction of
the baseline); a value must be worse than the baseline by more than the
allowed slack to count as a regression.
"""

from __future__ import annotations

import dataclasses
import json
import os
import re
import subprocess
import uuid
from datetime import datetime, timezone
from pathlib import Path
from typing import Optional

from .config import find_metrics_root


_DIRECTIONS = {"higher_is_better", "lower_is_better", "neutral"}

# Process-wide run id so every metric recorded in one pytest run shares it.
_RUN_ID: Optional[str] = None


def _now_iso() -> str:
    return datetime.now(timezone.utc).isoformat(timespec="seconds")


def run_id() -> str:
    """Return a stable id for the current process/run (or ``EVALPILOT_RUN_ID``)."""
    global _RUN_ID
    if _RUN_ID is None:
        _RUN_ID = os.environ.get("EVALPILOT_RUN_ID") or (
            datetime.now(timezone.utc).strftime("%Y%m%dT%H%M%S") + "-" + uuid.uuid4().hex[:8]
        )
    return _RUN_ID


def git_sha(repo: Optional[Path] = None) -> Optional[str]:
    """Return the short git SHA of ``repo`` (cwd by default), or None."""
    try:
        out = subprocess.run(
            ["git", "rev-parse", "--short", "HEAD"],
            cwd=str(repo) if repo else None,
            capture_output=True,
            text=True,
            check=False,
        )
        sha = out.stdout.strip()
        return sha or None
    except (FileNotFoundError, OSError):
        return None


def metric_slug(eval_id: str, name: str) -> str:
    """Stable filesystem slug identifying a metric series."""
    raw = f"{eval_id}__{name}" if eval_id else name
    return re.sub(r"[^A-Za-z0-9._-]+", "_", raw).strip("_")


def history_path(eval_id: str, name: str,
                 metrics_root: Optional[Path] = None) -> Path:
    """Return the JSONL history path for a metric series."""
    root = metrics_root or find_metrics_root()
    return root / metric_slug(eval_id, name) / "history.jsonl"


@dataclasses.dataclass(frozen=True)
class MetricResult:
    """Outcome of recording one metric value, with its regression verdict."""

    eval_id: str
    name: str
    value: float
    unit: str
    direction: str
    baseline: Optional[float]
    baseline_strategy: str
    delta: Optional[float]
    pct_delta: Optional[float]
    regressed: bool
    tolerance: float
    tolerance_pct: float
    run_id: str
    git_sha: Optional[str]
    timestamp: str
    history_path: Path

    def as_record(self) -> dict:
        return {
            "ts": self.timestamp,
            "run_id": self.run_id,
            "git_sha": self.git_sha,
            "eval_id": self.eval_id,
            "name": self.name,
            "value": self.value,
            "unit": self.unit,
            "direction": self.direction,
            "baseline": self.baseline,
            "baseline_strategy": self.baseline_strategy,
            "delta": self.delta,
            "pct_delta": self.pct_delta,
            "regressed": self.regressed,
            "tolerance": self.tolerance,
            "tolerance_pct": self.tolerance_pct,
        }

    def assert_no_regression(self, *, log_path=None) -> None:
        """Raise ``AssertionError`` if this metric regressed (for gating tests)."""
        if not self.regressed:
            return
        log_hint = f"\n  log: {log_path}" if log_path else ""
        raise AssertionError(
            f"Metric {self.name!r} regressed: value={self.value} "
            f"baseline={self.baseline} ({self.baseline_strategy}, "
            f"{self.direction}); delta={self.delta} "
            f"tolerance={self.tolerance} tolerance_pct={self.tolerance_pct}."
            f"{log_hint}"
        )

    def summary(self) -> str:
        base = (
            f"{self.baseline} ({self.baseline_strategy})"
            if self.baseline is not None
            else "(no baseline yet)"
        )
        verdict = "REGRESSED" if self.regressed else "ok"
        return (
            f"metric {self.name}={self.value}{self.unit and ' ' + self.unit} "
            f"vs baseline {base} -> {verdict}"
        )


def load_history(eval_id: str, name: str,
                 metrics_root: Optional[Path] = None) -> list[dict]:
    """Return all prior records for a metric series (oldest first)."""
    path = history_path(eval_id, name, metrics_root)
    if not path.exists():
        return []
    records = []
    for line in path.read_text(encoding="utf-8").splitlines():
        line = line.strip()
        if not line:
            continue
        try:
            records.append(json.loads(line))
        except json.JSONDecodeError:
            continue
    return records


def _resolve_baseline(history: list[dict], *, strategy: str, direction: str,
                      window: int, pinned: Optional[float]) -> Optional[float]:
    values = [r["value"] for r in history if isinstance(r.get("value"), (int, float))]
    if strategy == "pinned":
        return pinned
    if not values:
        return None
    if strategy == "last":
        return float(values[-1])
    if strategy == "rolling_mean":
        window_vals = values[-window:] if window > 0 else values
        return sum(window_vals) / len(window_vals)
    if strategy == "best":
        return max(values) if direction == "higher_is_better" else min(values)
    raise ValueError(f"Unknown baseline strategy: {strategy!r}")


def _is_regression(value: float, baseline: Optional[float], *, direction: str,
                   tolerance: float, tolerance_pct: float) -> bool:
    if baseline is None or direction == "neutral":
        return False
    slack = max(tolerance, abs(baseline) * tolerance_pct)
    if direction == "higher_is_better":
        # worse = lower; regression if value < baseline - slack
        return value < baseline - slack
    if direction == "lower_is_better":
        # worse = higher; regression if value > baseline + slack
        return value > baseline + slack
    return False


def record_metric(
    *,
    name: str,
    value: float,
    eval_id: str = "",
    direction: str = "higher_is_better",
    unit: str = "",
    baseline_strategy: str = "last",
    baseline: Optional[float] = None,
    window: int = 5,
    tolerance: float = 0.0,
    tolerance_pct: float = 0.0,
    metrics_root: Optional[Path] = None,
    repo: Optional[Path] = None,
) -> MetricResult:
    """Record ``value`` for a metric series and compute its regression verdict.

    Appends one JSON line to the series' ``history.jsonl`` and returns a
    :class:`MetricResult`. The comparison baseline is computed from the
    *prior* history (this value is excluded), so the very first recording
    has ``baseline=None`` and ``regressed=False``.
    """
    if direction not in _DIRECTIONS:
        raise ValueError(f"direction must be one of {_DIRECTIONS}, got {direction!r}")
    value = float(value)

    history = load_history(eval_id, name, metrics_root)
    resolved_baseline = _resolve_baseline(
        history, strategy=baseline_strategy, direction=direction,
        window=window, pinned=baseline,
    )
    delta = (value - resolved_baseline) if resolved_baseline is not None else None
    pct_delta = (
        (delta / resolved_baseline)
        if (delta is not None and resolved_baseline not in (None, 0))
        else None
    )
    regressed = _is_regression(
        value, resolved_baseline, direction=direction,
        tolerance=tolerance, tolerance_pct=tolerance_pct,
    )

    result = MetricResult(
        eval_id=eval_id,
        name=name,
        value=value,
        unit=unit,
        direction=direction,
        baseline=resolved_baseline,
        baseline_strategy=baseline_strategy,
        delta=delta,
        pct_delta=pct_delta,
        regressed=regressed,
        tolerance=tolerance,
        tolerance_pct=tolerance_pct,
        run_id=run_id(),
        git_sha=git_sha(repo),
        timestamp=_now_iso(),
        history_path=history_path(eval_id, name, metrics_root),
    )

    result.history_path.parent.mkdir(parents=True, exist_ok=True)
    with result.history_path.open("a", encoding="utf-8") as f:
        f.write(json.dumps(result.as_record()) + "\n")
    return result


def summarize(eval_id: str, name: str,
              metrics_root: Optional[Path] = None) -> dict:
    """Return a compact trend summary for a metric series."""
    history = load_history(eval_id, name, metrics_root)
    values = [r["value"] for r in history if isinstance(r.get("value"), (int, float))]
    return {
        "eval_id": eval_id,
        "name": name,
        "count": len(history),
        "first": values[0] if values else None,
        "latest": values[-1] if values else None,
        "min": min(values) if values else None,
        "max": max(values) if values else None,
        "mean": (sum(values) / len(values)) if values else None,
        "regressions": sum(1 for r in history if r.get("regressed")),
        "history_path": str(history_path(eval_id, name, metrics_root)),
    }


def iter_series(metrics_root: Optional[Path] = None):
    """Yield ``(slug, history_path)`` for every metric series on disk."""
    root = metrics_root or find_metrics_root()
    if not root.exists():
        return
    for series_dir in sorted(p for p in root.iterdir() if p.is_dir()):
        hp = series_dir / "history.jsonl"
        if hp.exists():
            yield series_dir.name, hp


__all__ = [
    "MetricResult",
    "record_metric",
    "load_history",
    "summarize",
    "history_path",
    "metric_slug",
    "iter_series",
    "run_id",
    "git_sha",
]

File: agent-packs/eval-pilot/engine-ts/src/metrics.ts
/**
 * Numeric metric results with JSONL history and trend/regression compare.
 *
 * Each recorded value is appended as one JSON line to a **committed** history
 * file: `<eval_root>/_metrics/<metric-slug>/history.jsonl`.
 *
 * Recording a metric also **compares it to a baseline** and flags regressions.
 * Baseline strategies: "last", "rolling_mean", "best", "pinned".
 * Directions: "higher_is_better", "lower_is_better", "neutral".
 */

import { spawnSync } from "node:child_process";
import {
  appendFileSync,
  existsSync,
  mkdirSync,
  readdirSync,
  readFileSync,
  statSync,
} from "node:fs";
import * as path from "node:path";
import { randomUUID } from "node:crypto";
import { findMetricsRoot } from "./config.js";

const DIRECTIONS = new Set(["higher_is_better", "lower_is_better", "neutral"]);

// Process-wide run id so every metric recorded in one run shares it.
let RUN_ID: string | null = null;

function nowIso(): string {
  return new Date().toISOString().replace(/\.\d{3}Z$/, "+00:00");
}

/** Return a stable id for the current process/run (or `EVALPILOT_RUN_ID`). */
export function runId(): string {
  if (RUN_ID === null) {
    const env = process.env.EVALPILOT_RUN_ID;
    if (env) {
      RUN_ID = env;
    } else {
      const now = new Date();
      const stamp =
        now.toISOString().replace(/[-:]/g, "").replace(/\.\d{3}Z$/, "");
      RUN_ID = `${stamp}-${randomUUID().replace(/-/g, "").slice(0, 8)}`;
    }
  }
  return RUN_ID;
}

/** Return the short git SHA of `repo` (cwd by default), or null. */
export function gitSha(repo?: string): string | null {
  try {
    const out = spawnSync("git", ["rev-parse", "--short", "HEAD"], {
      cwd: repo,
      encoding: "utf-8",
    });
    const sha = (out.stdout ?? "").trim();
    return sha || null;
  } catch {
    return null;
  }
}

/** Stable filesystem slug identifying a metric series. */
export function metricSlug(evalId: string, name: string): string {
  const raw = evalId ? `${evalId}__${name}` : name;
  return raw.replace(/[^A-Za-z0-9._-]+/g, "_").replace(/^_+|_+$/g, "");
}

/** Return the JSONL history path for a metric series. */
export function historyPath(
  evalId: string,
  name: string,
  metricsRoot?: string,
): string {
  const root = metricsRoot ?? findMetricsRoot();
  return path.join(root, metricSlug(evalId, name), "history.jsonl");
}

/** Outcome of recording one metric value, with its regression verdict. */
export interface MetricResult {
  eval_id: string;
  name: string;
  value: number;
  unit: string;
  direction: string;
  baseline: number | null;
  baseline_strategy: string;
  delta: number | null;
  pct_delta: number | null;
  regressed: boolean;
  tolerance: number;
  tolerance_pct: number;
  run_id: string;
  git_sha: string | null;
  timestamp: string;
  history_path: string;
}

export function metricAsRecord(m: MetricResult): Record<string, unknown> {
  return {
    ts: m.timestamp,
    run_id: m.run_id,
    git_sha: m.git_sha,
    eval_id: m.eval_id,
    name: m.name,
    value: m.value,
    unit: m.unit,
    direction: m.direction,
    baseline: m.baseline,
    baseline_strategy: m.baseline_strategy,
    delta: m.delta,
    pct_delta: m.pct_delta,
    regressed: m.regressed,
    tolerance: m.tolerance,
    tolerance_pct: m.tolerance_pct,
  };
}

/** Throw an Error if this metric regressed (for gating checks). */
export function assertNoRegression(
  m: MetricResult,
  opts: { logPath?: string } = {},
): void {
  if (!m.regressed) return;
  const logHint = opts.logPath ? `\n  log: ${opts.logPath}` : "";
  throw new Error(
    `Metric '${m.name}' regressed: value=${m.value} baseline=${m.baseline} ` +
      `(${m.baseline_strategy}, ${m.direction}); delta=${m.delta} ` +
      `tolerance=${m.tolerance} tolerance_pct=${m.tolerance_pct}.${logHint}`,
  );
}

export function metricSummary(m: MetricResult): string {
  const base =
    m.baseline !== null
      ? `${m.baseline} (${m.baseline_strategy})`
      : "(no baseline yet)";
  const verdict = m.regressed ? "REGRESSED" : "ok";
  const unitPart = m.unit ? ` ${m.unit}` : "";
  return `metric ${m.name}=${m.value}${unitPart} vs baseline ${base} -> ${verdict}`;
}

/** Return all prior records for a metric series (oldest first). */
export function loadHistory(
  evalId: string,
  name: string,
  metricsRoot?: string,
): Array<Record<string, any>> {
  const p = historyPath(evalId, name, metricsRoot);
  if (!existsSync(p)) return [];
  const records: Array<Record<string, any>> = [];
  for (const rawLine of readFileSync(p, "utf-8").split(/\r?\n/)) {
    const line = rawLine.trim();
    if (!line) continue;
    try {
      records.push(JSON.parse(line));
    } catch {
      continue;
    }
  }
  return records;
}

function numericValues(history: Array<Record<string, any>>): number[] {
  return history
    .map((r) => r.value)
    .filter((v): v is number => typeof v === "number");
}

function resolveBaseline(
  history: Array<Record<string, any>>,
  opts: {
    strategy: string;
    direction: string;
    window: number;
    pinned: number | null;
  },
): number | null {
  const values = numericValues(history);
  if (opts.strategy === "pinned") return opts.pinned;
  if (!values.length) return null;
  if (opts.strategy === "last") return Number(values[values.length - 1]);
  if (opts.strategy === "rolling_mean") {
    const windowVals = opts.window > 0 ? values.slice(-opts.window) : values;
    return windowVals.reduce((a, b) => a + b, 0) / windowVals.length;
  }
  if (opts.strategy === "best") {
    return opts.direction === "higher_is_better"
      ? Math.max(...values)
      : Math.min(...values);
  }
  throw new Error(`Unknown baseline strategy: ${JSON.stringify(opts.strategy)}`);
}

function isRegression(
  value: number,
  baseline: number | null,
  opts: { direction: string; tolerance: number; tolerancePct: number },
): boolean {
  if (baseline === null || opts.direction === "neutral") return false;
  const slack = Math.max(opts.tolerance, Math.abs(baseline) * opts.tolerancePct);
  if (opts.direction === "higher_is_better") {
    return value < baseline - slack;
  }
  if (opts.direction === "lower_is_better") {
    return value > baseline + slack;
  }
  return false;
}

export interface RecordMetricArgs {
  name: string;
  value: number;
  evalId?: string;
  direction?: string;
  unit?: string;
  baselineStrategy?: string;
  baseline?: number | null;
  window?: number;
  tolerance?: number;
  tolerancePct?: number;
  metricsRoot?: string;
  repo?: string;
}

/** Record `value` for a metric series and compute its regression verdict. */
export function recordMetric(args: RecordMetricArgs): MetricResult {
  const direction = args.direction ?? "higher_is_better";
  if (!DIRECTIONS.has(direction)) {
    throw new Error(
      `direction must be one of ${[...DIRECTIONS].join(", ")}, got ` +
        JSON.stringify(direction),
    );
  }
  const value = Number(args.value);
  const evalId = args.evalId ?? "";
  const name = args.name;
  const baselineStrategy = args.baselineStrategy ?? "last";
  const window = args.window ?? 5;
  const tolerance = args.tolerance ?? 0.0;
  const tolerancePct = args.tolerancePct ?? 0.0;

  const history = loadHistory(evalId, name, args.metricsRoot);
  const resolvedBaseline = resolveBaseline(history, {
    strategy: baselineStrategy,
    direction,
    window,
    pinned: args.baseline ?? null,
  });
  const delta = resolvedBaseline !== null ? value - resolvedBaseline : null;
  const pctDelta =
    delta !== null && resolvedBaseline !== null && resolvedBaseline !== 0
      ? delta / resolvedBaseline
      : null;
  const regressed = isRegression(value, resolvedBaseline, {
    direction,
    tolerance,
    tolerancePct,
  });

  const hp = historyPath(evalId, name, args.metricsRoot);
  const result: MetricResult = {
    eval_id: evalId,
    name,
    value,
    unit: args.unit ?? "",
    direction,
    baseline: resolvedBaseline,
    baseline_strategy: baselineStrategy,
    delta,
    pct_delta: pctDelta,
    regressed,
    tolerance,
    tolerance_pct: tolerancePct,
    run_id: runId(),
    git_sha: gitSha(args.repo),
    timestamp: nowIso(),
    history_path: hp,
  };

  mkdirSync(path.dirname(hp), { recursive: true });
  appendFileSync(hp, JSON.stringify(metricAsRecord(result)) + "\n", "utf-8");
  return result;
}

/** Return a compact trend summary for a metric series. */
export function summarize(
  evalId: string,
  name: string,
  metricsRoot?: string,
): Record<string, unknown> {
  const history = loadHistory(evalId, name, metricsRoot);
  const values = numericValues(history);
  return {
    eval_id: evalId,
    name,
    count: history.length,
    first: values.length ? values[0] : null,
    latest: values.length ? values[values.length - 1] : null,
    min: values.length ? Math.min(...values) : null,
    max: values.length ? Math.max(...values) : null,
    mean: values.length
      ? values.reduce((a, b) => a + b, 0) / values.length
      : null,
    regressions: history.filter((r) => r.regressed).length,
    history_path: historyPath(evalId, name, metricsRoot),
  };
}

/** Yield `[slug, historyPath]` for every metric series on disk. */
export function iterSeries(metricsRoot?: string): Array<[string, string]> {
  const root = metricsRoot ?? findMetricsRoot();
  if (!existsSync(root)) return [];
  const out: Array<[string, string]> = [];
  const dirs = readdirSync(root, { withFileTypes: true })
    .filter((e) => e.isDirectory())
    .map((e) => e.name)
    .sort();
  for (const name of dirs) {
    const hp = path.join(root, name, "history.jsonl");
    if (existsSync(hp) && statSync(hp).isFile()) out.push([name, hp]);
  }
  return out;
}
